### ISSUE TITLE
viewModel is not found by webpack

### ISSUE THREAD

**Issue by dkent600** [aurelia/dialog#298]
**I'm submitting a bug report**

* **Library Version:**
1.0.0-rc.1.0.3

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
6.10.3
* **NPM Version:**
5.0.0

* **JSPM OR Webpack AND Version**
webpack 2.5.1

* **Browser:**
All

* **Language:**
TypeScript 2.3

**Current behavior:**
DialogService is not finding my viewModel.  It does not report the problem and instead puts up an entirely empty and invisible dialog.

I invoke `.open`  like this:

```
        return this.dialogService.open(settings) ...
```

where `settings.viewModel` is set like this:
```
import {Prompt} from "../resources/commonDialogs/prompt";
.
.
.
settings.viewModel = Prompt;

```

(Note: `Prompt` is my own, not from the Aurelia-dialog example.)

DialogService sees  the `Prompt` viewModel function and tries to do something with it here:

```
 DialogService.prototype.ensureViewModel = function (compositionContext) {
        if (typeof compositionContext.viewModel === 'function') {
            compositionContext.viewModel = __WEBPACK_IMPORTED_MODULE_1_aurelia_metadata__["Origin"].get(compositionContext.viewModel).moduleId;
        }
        if (typeof compositionContext.viewModel === 'string') {
            return this.compositionEngine.ensureViewModel(compositionContext);
        }
        return Promise.resolve(compositionContext);
    };
```

But `WEBPACK_IMPORTED_MODULE_1_aurelia_metadata__["Origin"].get(compositionContext.viewModel).moduleId` returns `undefined`.

All of the relevant code is contained in a vendor.js file generated by webpack.

**Expected/desired behavior:**

* **What is the expected behavior?**
Should find my viewmodel or else report an error.  If there are additional steps that I'm not following, then they should be documented.

* **What is the motivation / use case for changing the behavior?**
Supplying a viewModel in the settings is supposed to work.  I assume webpack is part of the use case as well.

**Comment by dkent600** [comment]
Would be nice to know if there are any workarounds for the time being.  At the moment I've got no dialogs...

**Comment by StrahilKazlachev** [comment]
OK, finally found it - a nice [explanation](https://github.com/aurelia/dialog/issues/283#issuecomment-298885151) of what is happening. It is related to a separate issue but the `import` part is the same.

**Comment by dkent600** [comment]
@StrahilKazlachev Your suggestion,

`settings.viewModel = PLATFORM.moduleName('../resources/commonDialogs/prompt')`

doesn't work.  It results in the following error message:

"Unable to find module with ID: ../resources/commonDialogs/Prompt"

I haven't had a chance to experiment to figure out what *does* work.

**Comment by dkent600** [comment]
Following my previous comment above ^^^ , I can make the error message go away by referencing the `Prompt` viewmodel in the code to force webpack to include it in the bundle.

However, when I do that, while there is no more error message, there is also no dialog being shown.  The `aurelia-webpack-loader` creates a promise to load the module, but the promise never resolves, so nothing ever happens.

(At this point my project is dead in the water until I can get a solution for this.)

**Comment by StrahilKazlachev** [comment]
@dkent600
https://github.com/aurelia/dialog/issues/283#issuecomment-298895875
https://github.com/aurelia/dialog/issues/283#issuecomment-299110588

**Comment by dkent600** [comment]
@StrahilKazlachev

Thanks.  As far as I can tell, https://github.com/aurelia/dialog/issues/283#issuecomment-298895875 is the work-around that you originally proposed that I already found doesn't work (see above).

#283 (comment)  Is a no-op  (I get the undefined viewemodel as returned by `DialogService.prototype.ensureViewModel`, as I described above).

You are taking these work-arounds from an issue that is related to this one, but is actually a quite different use case, and that may be why the workarounds aren't working for me.

**Comment by StrahilKazlachev** [comment]
I don't see what more I can do without a repro.

**Comment by dkent600** [comment]
Do you have a GitHub repository set up that I could fork and use to make a gist repo for you?

**Comment by StrahilKazlachev** [comment]
No. Also in this case we also want to see your webpack setup, since this scenario was tested and worked - with the caveat of needing `PLATFORM.moduleName`.

**Comment by dkent600** [comment]
Yes, you would get the webpack setup with the fork

**Comment by dkent600** [comment]
OK, here is a repro.  To run it:
1)  unzip it
2) on a Windows command prompt, run the following commands:
    a) `dotnet restore`
    b) `npm install`
    c) `dotnet run`
3) browse to http://localhost:5000
4) Click on "Open local dialog" button and see that it works as advertised with no workarounds required
5) Click on "Open plugin dialog" button and see that it doesn't work.  The code is using the recommended workaround.  In the javascript console you'll see an error message indicating that it can't find the viewmodel, even though it is present in the vendor.js bundle.

Let me know if you  have any problems running the app.  Thanks.

[Aurelia-Dialog-Test.zip](https://github.com/aurelia/dialog/files/1081593/Aurelia-Dialog-Test.zip)

**Comment by jods4** [comment]
Disclaimer: I haven't read most of the thread above, sorry.

Wanted to say: `PLATFORM.moduleName('../resources/commonDialogs/prompt')` is most likely not good because at runtime `aurelia-loader` is going to try to resolve the relative path `../resources/commonDialogs/prompts` and situations where it actually works are few.

If this is your own app I suggest you always use absolute module ids, like `moduleName('resources/commonDialogs/prompt')`. Of course you may need to adjust this absolute path based on where your application root.

Nobody asked which Aurelia Webpack plugin you use. `moduleName` assumes that you use the new 2.0 RC2 plugin, otherwise it will have absolutely no effect on the 1.x versions.

**Comment by dkent600** [comment]
@jods4 Thanks for the suggestion.  I can only find one other path that will get past compilation:  

`moduleName("ferngully-aurelia-tools/dist/native-modules/resources/commonDialogs/prompt")`.

When I use this it still fails at runtime, but with a different error message.  Instead of saying it can't find "prompt" it says it can't find "prompt.html".

In any case, even if it worked, I would not be wild about this work-around because it hardcodes the `native-modules` module type.

**Comment by jods4** [comment]
ah it's inside a library? isn't great then, indeed. That's why I said "if this is your own app".
`ferngully-aurelia-tools/resources/commonDialogs/prompt` is worth a shot (there's a plugin inside AureliaPlugin that allows that kind of things) but I'm not sure it'll work.

What's the full path it says it can't find for the `prompt.html` part? Is that HTML view expected -- I'm not too familiar with `aurelia-dialog`?

**Comment by dkent600** [comment]
@jods4  OK, looks like we have a workaround!  `PLATFORM.moduleName("ferngully-aurelia-tools/resources/commonDialogs/prompt")` both compiled, functioned properly and is acceptable.

That gets me past this issue.  Thanks!

**Comment by dkent600** [comment]
If anyone would like, I can provide another repro for the case where we pass class name and it fails.

**Comment by dkent600** [comment]
If anyone would like, I can provide another repro for the case where we pass class name and it fails.

In any case, I'm resubmitting the original zip to lock-in the version of the plugin "library" that it grabs.

**Comment by dkent600** [comment]
If anyone would like, I can provide another repro for the case where we pass the class name and it fails.

In any case, I'm resubmitting the original zip to lock-in the version of the plugin "library" that it grabs.

**Comment by Vheissu** [comment]
@dkent600 It sounds like [this link](https://github.com/aurelia/webpack-plugin/wiki/Managing-dependencies#moduledependenciesplugin) might be what you're after (leading off of @jods4 point). The issue is modules that do not define their dependencies, so Webpack has no idea they exist unless you specifically tell it about them.

**Comment by dkent600** [comment]
@Vheissu That is definitely good to know about, thanks!  

However, in my case, I am the plugin author so `ModuleDependenciesPlugin` is not required.  What was required for me, as a plugin author, is that when I supply a value for the `DialogSettings.viewModel`, I must supply it as follows:

 ```
settings.viewModel = PLATFORM.moduleName("[absolutePathToView]");
```

...where the path starts with the name of the plugin.

Example that works for my case:
```
settings.viewModel = PLATFORM.moduleName('ferngully-aurelia-tools/resources/commonDialogs/prompt');
```

Neither of the following alternatives will work:

1) relative paths, for example: `settings.viewModel = PLATFORM.moduleName('../resources/commonDialogs/prompt')`.  This *should* work, but doesn't.
2) the class itself, for example: `settings.viewModel = Prompt`.  This works in an application but not in a plugin.

**Comment by dkent600** [comment]
@Vheissu @StrahilKazlachev

I think it would be good if all this ( just above ^^^) were documented [here](https://github.com/aurelia/webpack-plugin/wiki/Managing-dependencies#platformmodulename).

Namely:

1) include `aurelia-dialog`-in-a-plugin in the list of cases where using `PLATFORM.moduleName` is required (would be good if this were also documented in the `Aurelia-dialog` documentation for `DialogSettings.viewModel`).
2) describe when using a full path to the module is required

**Comment by jods4** [comment]
I thought `aurelia-dialog` did not require `moduleName` usage, even for models loaded through ES imports? I thought that we did something to fix the origin detection code in that case, didn't we?

Is this a general problem or is it linked to the fact that it's inside a library? 

I'm sorry I didn't follow everything in the thread and I don't know the inner workings of `aurelia-dialog` well. I can add stuff to the docs no problem, but it needs to be a clear explanation.

**Comment by dkent600** [comment]
@jods4 `moduleName` is currently only required inside a plugin.

**Comment by jods4** [comment]
ok, did you figure out why? 

Is there something we could do to make it work?
If not, can I have an explanation why it doesn't work in a plugin? I can add that to the docs.

**Comment by dkent600** [comment]
No, I didn't figure it out, though I gave some hints farther above in this issue report.  But I can easily provide a repro that will enable someone to figure it out.  Let me know...

**Comment by jods4** [comment]
@StrahilKazlachev @PWKad any insight?

**Comment by PWKad** [comment]
From what I can tell here @dkent600 is a plugin author trying to re-expose the dialog plugin functionality through a plugin to an app being bundled by WebPack.  That nor anything mentioned above appears to be related to any issues in dialog so I'm closing this issue for now.

Doug - If you feel there is documentation needs for Webpack plugin setup please feel free to open or track a documentation issue there.

**Comment by dkent600** [comment]
@jods4 Looks like the ball is being passed to you to  determine whether or not this is to be viewed as a bug or a feature, and whose bug it is.  Please let me know if I can assist on that (such as providing a repro).

As for my request for documentation, I believe it will depend on whether this will be fixed or will require the work-around of using `moduleName`.

**Comment by PWKad** [comment]
@dkent600  PR's are welcome for all documentation issues

**Comment by dkent600** [comment]
@PWKad it is not yet clear what is going to be the resolution of this, so I wouldn't know, at this point, what to document.

**Comment by StrahilKazlachev** [comment]
@dkent600 How can I rebuild the repro? When I run `webpack` it fails with a lot of errors:
```
    ERROR in Aurelia-Dialog-Test\node_modules\@types\jquery\index.d.ts
    (4101,25): error TS2314: Generic type 'PlainObject<T, any>' requires 2 type argument(s).

    ERROR in Aurelia-Dialog-Test\node_modules\@types\jquery\index.d.ts
    (4105,43): error TS2314: Generic type 'Promise<TResolve, TReject, any, TNotify>' requires 4 type argument(s).

    ERROR in Aurelia-Dialog-Test\node_modules\@types\jquery\index.d.ts
    (4151,48): error TS2531: Object is possibly 'null'.

    ERROR in Aurelia-Dialog-Test\node_modules\@types\jquery\index.d.ts
    (4158,24): error TS2304: Cannot find name 'TTarget'.

    ERROR in Aurelia-Dialog-Test\node_modules\@types\jquery\index.d.ts
    (4165,15): error TS2304: Cannot find name 'TData'.
```

**Comment by dkent600** [comment]
@StrahilKazlachev 

Running `webpack` isn't part of the instructions I gave, here https://github.com/aurelia/dialog/issues/298#issuecomment-309095582

Let me know if those don't work.

**Comment by dkent600** [comment]
To rebuild the webpack vendor bundle, do this:

`node node_modules/webpack/bin/webpack.js --config webpack.config.vendor.js`

The app bundle is automatically rebuilt when you start the app, but you can do it manually like this:

`node node_modules/webpack/bin/webpack.js`

**Comment by dkent600** [comment]
FYI, this is the standard template produced for JavaScriptServices: https://github.com/aspnet/JavaScriptServices.

**Comment by StrahilKazlachev** [comment]
@dkent600 I ran it according your instructions. Now I'm asking how do I rebuild `wwwroot/dist`? I want to change some things.

**Comment by StrahilKazlachev** [comment]
OK, so had to launch it from VS to get the `dist` rebuid. I'll post more tomorrow.

**Comment by dkent600** [comment]
You don't technically have to relaunch from VS.  `dotnet run` should work the same.

**Comment by dkent600** [comment]
It just occurred to me that if running via `dotnet run` you may actually need to do it like this:  `dotnet run environment=development`

**Comment by StrahilKazlachev** [comment]
@dkent600 This sample just doesn't work. Following just your normal run steps(restore, install, run) it's OK.
After I rebuild the vendors bundle - `node node_modules/webpack/bin/webpack.js --config webpack.config.vendor.js`, I get `Uncaught ReferenceError: vendor_aeffab828c1031bcccd2 is not defined`.
Then I rebuilt the app bundle - `node node_modules/webpack/bin/webpack.js`(a *lot* of TS errors), and now the error is `Uncaught (in promise) Error: Unable to find module with ID: aurelia-framework`.

**Comment by jods4** [comment]
Quoting @dkent600 
> it is not yet clear what is going to be the resolution of this, so I wouldn't know, at this point, what to document.

It is not clear for me either what the core of this issue is... A _minimal_ demo or a good explanation would help. 

I am currently preparing a Webpack plugin RC3 release (which will hopefully be the last one before 2.0) and I really don't mind including a tweak of it helps some specific scenarios. I'd also gladly enhance the docs where it might be needed, I'm going to add a few new pages for RC3 anyway.

**Comment by dkent600** [comment]
@StrahilKazlachev @jods4 I will  check out this problem with building the bundles in the next few hours.

**Comment by StrahilKazlachev** [comment]
It's already late for me, but I was thinking that the next thing I can try is to setup a repro using one of jods4 demos and only focus on to what ` Origin.get().moduleId` is resolved in a couple of scenarios.

**Comment by dkent600** [comment]
@StrahilKazlachev  I just unzipped the second zip that I posted [here](https://github.com/aurelia/dialog/issues/298#issuecomment-309147197) and was able to rebuild both the vendor and app bundles without problem, and the app ran fine with them.

Are you sure you used the second zip?  If so, I guess I would need more info or be standing over your shoulder.

**Comment by StrahilKazlachev** [comment]
@dkent600 Could you try the following:
- in your plugin > in `services/dialog-service`
  - `import {Prompt} from '../resources/commonDialogs/prompt';`
  - `settings.viewModel = Prompt;`
- in your sample/test/repro app > in `main`(app entry point)
  - in addition to `.plugin(PLATFORM.moduleName("aurelia-dialog"))` *add also* `.plugin(PLATFORM.moduleName('ferngully-aurelia-tools'))`

**Comment by dkent600** [comment]
@StrahilKazlachev That case fails with:

> vendor.js?v=DWMVnLLAWpT78HMhLZZayWHtZL3fehw_Fz93D5c1YVk:67829
>  Uncaught (in promise) TypeError: Cannot read property 'attached' of null
    at DialogRenderer.attach (vendor.js?v=DWMVnLLAWpT78HMhLZZayWHtZL3fehw_Fz93D5c1YVk:67829)
    at DialogRenderer.showDialog (vendor.js?v=DWMVnLLAWpT78HMhLZZayWHtZL3fehw_Fz93D5c1YVk:67901)
    at vendor.js?v=DWMVnLLAWpT78HMhLZZayWHtZL3fehw_Fz93D5c1YVk:68019
    at <anonymous>

The line of code:

`dialogController.controller.attached();`

**Comment by dkent600** [comment]
BTW, with regard to rebuilding the repro app:

The tools I'm using:
npm v5.0.3
tsc v2.3.4
node v8.1.2
webpack v2.6.1

**Comment by StrahilKazlachev** [comment]
@dkent600 in the repro you've attached `.plugin(PLATFORM.moduleName('ferngully-aurelia-tools'))` is missing. So could you check you have it now?
In regard to the above error *if* you are passing a config callback to `aurelia-dialog` plugin, please *do call* `.useDefaults()` on the config object.
Do *not* rely on HMR - there might be something with it and the dialog.

**Comment by dkent600** [comment]
```
aurelia.use.standardConfiguration()
        .plugin(PLATFORM.moduleName("aurelia-dialog"))
        .plugin(PLATFORM.moduleName('ferngully-aurelia-tools'));
```

This is the original case that I reported in this issue.  I gave details that you can review above about what is happening.

**Comment by StrahilKazlachev** [comment]
@jods4 
1. Using your `01-No_splits` demo I opened  the prompt from `ferngully-aurelia-tools` successfully - the modification I did in the `ferngully-aurelia-tools` plugin was:
    * in services/dialog-service
      ```js
      import {Prompt} from '../resources/commonDialogs/prompt';
      // ...
      settings.viewModel = Prompt;
      ```
     * removed all PLATFORM.moduleName related to the prompt - in the plugin
     * and it just works, `Origin.get(compositionContext.viewModel).moduleId` is resolved OK.
2. In the provided repro with the same modification to the plugin `Origin.get(compositionContext.viewModel).moduleId` resolves to `undefined`.
The only difference I see is in the build setup - one bundle vs multiple bundles. The vendors bundle, in which is `ferngully-aurelia-tools`, is build with the DLL plugin. How can this affect your *magic* for `Origin.get(...).moduleId` you will say.

**Comment by jods4** [comment]
OK, now we're talking.
It's @niieani's magic, but thankfully I understand that part well enough that I can explain.

Aurelia has a lot of reliance on the _origin_ (module name) of resources (viewmodels & co.). As long as modules are loaded through `aurelia-loader`, they are tagged so that this works (it's done inside `aurelia-loader-webpack`).

The issue arises when something that was _not_ loaded through `aurelia-loader` is suddenly asked for origin by Aurelia. This happens very rarely, and the pattern above (`import`ed VM for `aurelia-dialog`) is the main occurrence of this.

Before I even got started on the Webpack 2.0 plugin, @niieani had identified this issue and came up with the following solution: he would look at every loaded module for an export that matches the object whose origin is asked for. 
This is done here: 
https://github.com/aurelia/metadata/blob/master/src/origin.js#L37-L56

As you can see it relies on `PLATFORM.eachModule`. How do you iterate all modules? Well, it's not part of the official API (it grabs directly at internals), but `aurelia-loader-webpack` provides this implementation:
https://github.com/aurelia/loader-webpack/blob/master/src/aurelia-loader-webpack.ts#L90-L100

And this is where the limitation lies. At the time when @niieani did this we did not have DLL support, which I added later. If you have some familiarity with Webpack internals you understand that the code I linked above is only looking at modules _inside the entry bundle._ Any other bundle (DLL... and maybe code splits, I would need to double-check that) is not examined.

This is the problem you've identified.

The root cause is in `aurelia-loader-webpack`, the solution would be to dig deeper into Webpack's internals and look at modules into other (e.g. DLL) bundles that have been loaded. As this is all undocumented stuff, it would need a bit of research.

**Comment by dkent600** [comment]
Thanks for working this out guys!

**Comment by jods4** [comment]
@dkent600 I'll try to fix it.

**Comment by jods4** [comment]
@StrahilKazlachev please re-open and assign this to me.

**Comment by PWKad** [comment]
Doesn't this issue belong in `aurelia-loader-webpack`?

**Comment by jods4** [comment]
I've done some tests and it works better than I though -- the origin is the proxy in main bundle rather than the real origin in the dll bundle... but I don't think that will make much difference and it's certainly not `undefined`.

How did you export the class in the DLL?
How did you import it on the other side?
Then `Origin.get()` is null?
What module-style are you using in the compiled code feeded to Webpack (ES, commonjs, amd)?

**Comment by StrahilKazlachev** [comment]
@jods4 Could you also check the repro, just in case? 
- From what I see, for the vendors bundle(DLL), the `native-modules` builds are fed to Webpack. Not sure for the app bundle.
- Inside `ferngully-aurelia-tools` - the `Prompt` `moduleId` is what we want to get.
  - *no* `PLATFORM.moduleName` is used as I said in my last comment.
  - in `resources/commonDialogs/prompt` - `export { Prompt };`
  - in `services/DialogService` - `import {Prompt} from '../resources/commonDialogs/prompt';`
  - both export and import happen in this plugin - which is bundled in the DLL bundle.
- `Origin.get()` is `{moduleId: undefined, moduleMember: undefined}`(we read the `moduleId`), when called inside `aurelia-dialog/dialog-service` for the passed in `Prompt`. In the vendor bundle this is - `compositionContext.viewModel = __WEBPACK_IMPORTED_MODULE_1_aurelia_metadata__["Origin"].get(compositionContext.viewModel).moduleId;`
- the modification I did to the repro
  -  in `node_modules/ferngully-aurelia-tools/dist/native-modules/services/dialog-service`
    ```js
    // added
    import {Prompt} from '../resources/commonDialogs/prompt';
    // ...
    // replaced
    settings.viewModel = PLATFORM.moduleName('ferngully-aurelia-tools/resources/commonDialogs/prompt');
    // with 
    settings.viewModel = Prompt;
    ```

**Comment by jods4** [comment]
OK I get it:
> both export and import happen in this plugin - which is bundled in the DLL bundle.

I imported in the main bundle, which "kind of" works. Because what you use from the DLL is proxied by a module in the main bundle, you find that module name. It's not "the real deal" but it shouldn't matter.

What is not used outside the DLL does not leak out. There's a complicated way, which involves emitting additional infos at compile-time in JS code that I could reach out that DLL modules cache. If you use DLLs transitively it's even more complicated -- or not going to work.

Thinking of what we're trying to achieve, I'm not sure it's the right way to go. 
Because in the end, your module id (origin) will be something like a number `5`.
Without knowing from which DLL it comes from, that's not very helpful. `aurelia-loader` always gets modules from the main bundle (it uses proxies for what is in other bundles) so requiring `5` won't give you the correct module!
Aurelia uses the origin to normalize the name of relative resources in some situations, which won't give anything good starting with `5` anyway.

Bottom line: I don't think adding tons of complexity to make `Origin.get()` "work" in this situation is worth it.

I think:
- we could tweak the origin logic to return `-1` for modules whose origin is unknown. It's as good as any other Webpack module id. As long as you don't really need the origin -- which is most of the time, including the use case from this issue.
- if you need something that has a reliable, working origin the user should have a `PLATFORM.moduleName` somewhere, even inside that module. If that's common enough maybe we can think of another API for that use case, like `@aurelia-resource` on top of the class.

What do you think?

**Comment by StrahilKazlachev** [comment]
@jods4 I think for now we should document such cases as *limitations* and how to deal with them.
What I'm not sure is *whether*, and if *how*, `PLATFORM.moduleName` should have runtime semantics? Like if invoked *inside* a *plugin* with a relative path(only if the moduleId is static), at runtime it should return the path relative to the app root? For the current case we **could** *have* recommended going with `{viewModel: PLATFORM.moduleName('../resources/commonDialogs/prompt')}`.
Now the **recommendation** is to do `{viewModel: PLATFORM.moduleName('ferngully-aurelia-tools/resources/commonDialogs/prompt')}`.

**Comment by jods4** [comment]
@StrahilKazlachev 
At one point I wondered if Webpack should replace the `moduleName` arg with a normalized one...
It's a double-edged sword:
- it would help people for sure (albeit not in this specific case, as there is no `moduleName` and that's the problem);
- but it would create a surprising behavior unlike other platform, so new problems for sure.

As far as this issue is concerned, I'm going to make a PR to return `"unknown_webpack_origin"` for unknown webpack origins rather than crashing. That will make more cases work (such as this one) although it's not a silver bullet. For the (rare) rest I don't see a solution better than adding a `moduleName` where not strictly necessary at first glance.

**Comment by jods4** [comment]
News: this just got a little worse.

If you use ES `import` to import your ViewModel, unless it came from a DLL the origin detection code usually works.

With Webpack 3, if you use `ModuleConcatenationPlugin` that safety net is very likely to break. Your ViewModel is likely to get scope merged with the module that uses it and then there's no way to make Origin work anymore.

I think we should discourage the ES `import` way, at least when using Webpack. There's no way to make it work with those scenarios. 
We need to clearly document that with Webpack you should `moduleName` your ViewModel.

**Comment by dkent600** [comment]
Thanks @jods4.  I would like to confirm that I understand what you are recommending here.

When you say 

> "I think we should discourage the ES import way, at least when using Webpack", 

are you referring only to the use case originally described in the original description of this issue?  Or are you speaking more broadly?  One reason I wonder is because I thought we had already agreed that using `moduleName` was going to be required in this scenario (and that this needed to be documented).  Perhaps all you are doing is further reinforcing that conclusion.

**Comment by jods4** [comment]
@dkent600 
I meant ES `import` a viewmodel that is then feeded as a ctor to `aurelia-dialog`.

With Webpack it creates unsolvable origin problems when using DLL and now when using `ModuleConcatenationPlugin`.

Instead you should pass the module name, wrapped with `moduleName`.

Using ES `import` in your code is fine. In fact it's the module system I recommend when using Webpack to bundle your app (it supports advanced optimizations such as `ModuleConcatenationPlugin`, tree shaking or exports renaming).

**Comment by dkent600** [comment]
@jods4  Ah, gotcha, so the idea is that this is a problem with `aurelia-dialog` not only when using a DLL, but also anytime when using `ModuleConcatenationPlugin`.  

Perhaps this might need to be documented in `aurelia-dialog` as well.

**Comment by jods4** [comment]
Not anytime, but very likely. 
If the module that contains your class is added to a concatenated module, then it won't work.

**Comment by ajoslin103** [comment]
I'm using the new aurelia-cli with webpack 

  * I followed the instructions on the aurelia-dialog page
  * I changed my style prefixes from ai- to ux-
  * I specified the viewModel using the PAL

I did not need to do any import of my dialog modelView
   I only needed to specify the viewModel properly 

I was previously specifying a relative path: '../../resources/dialogs/enrollment'
And I tried specifying an absolute path: '/resources/dialogs/enrollment'
Then I specified the path to the PAL -- relative to the src folder

and it worked: 

    this.dialogService
      .open({
        viewModel: PLATFORM.moduleName('resources/dialogs/enrollment'),
        position: (modalContainer) => {
          modalContainer.classList.add('enrollment_modal-container');
        }
      })
      .then(dialogResult => {
        Console.debug('dialogResult', dialogResult);
      })
      .catch(ThenError);

**Comment by StrahilKazlachev** [comment]
Added info about this in the docs #324. I don't see what more we can do here.
